### PR TITLE
raise exception if steps was defined but not used

### PR DESCRIPTION
**Pull request by trifonovmixail**
It looks like
```
self = <pytest_allure_dsl.AllureDSL object at 0x7f21bb81d198>

    def _check_steps_was_used(self):
        not_used_steps = []
    
        for key in self.steps.keys():
            if key not in self._steps_was_used:
                not_used_steps.append(key)
    
        if not_used_steps:
>           raise StepsWasNotUsed(*not_used_steps)
E           pytest_allure_dsl.StepsWasNotUsed: (5, 6)

pytest_allure_dsl.py:215: StepsWasNotUsed
```

### CODE REVIEW

File: pytest_allure_dsl.py [hunk 1]
```diff
@@ -198,6 +204,16 @@ def _yaml_load(self, string):
...
 
+    def _check_steps_was_used(self):
+        not_used_steps = []
+
+        for key in self.steps.keys():
+            if key not in self._steps_was_used:
+                not_used_steps.append(key)
+
+        if not_used_steps:
+            raise StepsWasNotUsed(*not_used_steps)
```
Comment (griddic) [inline pytest_allure_dsl.py:211]: I think it is helpful to describe some actions made by fixtures in the description of a test. 
So, I think we should afford to have steps described, but not implemented.

**Review by griddic — CHANGES_REQUESTED** [review]

File: pytest_allure_dsl.py (hunk 1 above)
Comment (trifonovmixail) [reply pytest_allure_dsl.py:211]: We can extand DSL as we wish for this
```
steps:
  1:
    title: some title of implemented step
    not-implemented-steps:
      1.1: some title of not implemented step 
````

or

```
steps:
  1: some title of implemented step
not-implemented-steps:
  1: some title of not implemented step
```

but if code will changed and doc string not, we don't know about this

File: pytest_allure_dsl.py [hunk 2]
```diff
@@ -275,6 +291,9 @@ def step(self, key, **kwargs):
         else:
             step_name = step
 
+        if key not in self._steps_was_used:
```
Comment (griddic) [inline pytest_allure_dsl.py]: You will not need this if statement if  _steps_was_used is a set.

**Review by griddic — APPROVED** [review]

File: pytest_allure_dsl.py (hunk 2 above)
Comment (trifonovmixail) [reply pytest_allure_dsl.py]: Yes, but set requires more memory :)
```
In [11]: set().__sizeof__()                                                                                                                                                                                                                  
Out[11]: 200

In [12]: list().__sizeof__()                                                                                                                                                                                                                 
Out[12]: 40
```

and ..

```
~/test$ time python3.6 -c "33 in list(range(10000))" 

real    0m0,026s
user    0m0,022s
sys     0m0,004s
~/test$ time python3.6 -c "33 in set(range(10000))" 

real    0m0,059s
user    0m0,049s
sys     0m0,010s
```

File: pytest_allure_dsl.py [hunk 3]
```diff
@@ -198,6 +204,16 @@ def _yaml_load(self, string):
             self.description_load_error = e
             return {}
 
+    def _check_steps_was_used(self):
```
Comment (igorji) [inline pytest_allure_dsl.py:207]: may be shorter:

```python
def _check_steps_was_used(self):

    not_used_steps = self.steps.keys() - self._steps_was_used

    if not_used_steps:
        raise StepsWasNotUsed(not_used_steps)
```

Comment (trifonovmixail) [reply pytest_allure_dsl.py:207]: How we know about not implemented steps? What we must delete from steps? pdb? breakpoint?

Comment (trifonovmixail) [reply pytest_allure_dsl.py:207]: Oh, sorry, i think len(...) - len(...) :) I can change it if you wish, but this is overhead

Comment (igorozt) [reply pytest_allure_dsl.py]: Not implemented steps will be passed to exception, if any

Comment (igorji) [reply pytest_allure_dsl.py:207]: As you wish :)

Comment (trifonovmixail) [reply pytest_allure_dsl.py:207]: @griddic ?

Comment (igorji) [reply pytest_allure_dsl.py:207]: upd: more clean version, see edited first comment

Comment (trifonovmixail) [reply pytest_allure_dsl.py:207]: @igorozt It's not working for python2, but supporting for this lib, TypeError will raised

Comment (igorji) [reply pytest_allure_dsl.py:207]: will work both for python 2 and 3:

`not_used_steps = set(self.steps.keys()) - set(self._steps_was_used)`

but consider https://pythonclock.org :)

Comment (trifonovmixail) [reply pytest_allure_dsl.py:207]: Yep! )

Comment (griddic) [reply pytest_allure_dsl.py:207]: > @griddic ?

I still prefer _steps_was_used to be a set )

Comment (trifonovmixail) [reply pytest_allure_dsl.py:207]: @griddic @igorozt I fixed as you wish :)

Comment (igorji) [reply pytest_allure_dsl.py:207]: Looks nice, except for tuple(..) which seems unnecessary

**Pull request merged**